text/0083-rawkv-cross-cluster-replication.md

```diff
@@ -0,0 +1,183 @@
+* And with the following acceptable loss:
+  * More read latency in tens of microseconds.
+  * A little more write overhead caused by [MCE Padding](https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format).
+  * An extra Garbage Collection is required (used with TxnKV).
+  * Possible more space occupied (related to update pattern).
+
+### 5. Incremental Scan
+
+On replication task resumed after paused, we scan *KVDB* to get the change data during pause by extracting timestamp encoded in key. The scan will be heavy as we actually scan all the keys, so we should avoid scanning as possible, and limit scan rate.
+
+*Incremental Scan in SQL / TxnKV scenario scans write_cf to get change data scope, which is much cheaper. So an alternative to reduce cost for RawKV is writing timestamp to write_cf other than encoded in key, but will introduce an additional write for each put request. We tend to the lower delay of writing now.*
+
+### 6. TiKV-CDC
+
+**TiKV-CDC** is a fork of TiCDC. The following modifications will be applied:
+
+* Owner:
+  
+  * A new type of [*changefeed*](https://docs.pingcap.com/tidb/stable/manage-ticdc#manage-replication-tasks-changefeed) which is specified by key range.
+  
+  * A new type of **task** which is specified as a **sub-range** of the whole key range, and is handled by a *KV Processor* on a [*Capture*](https://docs.pingcap.com/tidb/stable/ticdc-overview#ticdc-architecture).
+
+* Scheduler: A new scheduler which splits the whole key range of *changefeed* into *sub-ranges*, and distributes to *captures*.
+  
+  * For simplicity, we split whole key range into fixed number of parts, which can be specified by *changefeed* argument, and defaults to **2**. As we sink data to downstream TiKV by **batch put** and bottleneck of replication should not be on TiKV-CDC, a parallel of **2** will be enough.
+  
+  * Key range splitting is align with boundary of Regions. For simplicity, each *sub-range* contains almost equal number of Regions.
+  
+  * *Rebalance* can only be performed by *tikv-cdc cli*. As *rebalance* will trigger *incremental scan* and impact performance of TiKV, it is not running automatically. We need more information to determine the best moment for *rebalance* by TiKV-CDC itself in the future.
+
+* Capture: Accepts key-range tasks, and creates *KV Processors* to handle them.
+
+* Processor:
+  
+  * A new *KV Processor* which accepts key-range tasks, and processes key-value data
+
+* Sink:
+  
+  * A new *TiKV sink* to batch put data into downstream TiKV cluster, with configurable concurrency, batch size, and backoff parameters.
+
+## Prototype
+
+We have developed a [prototype](https://github.com/pingyu/tikv/issues/1) to verify feasibility of this proposal. All of the correctness validation cases are passed. And the benchmark results are as follows:
+
+| Case                          | gPRC P99 Duration (ms) | gPRC Avg Duration (ms) |
```

Comment:
should we update the test result based on encoding timestamp into the key part?

Reply:
Ah yes. We will update the result when we finish the benchmark this week.